staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -468,7 +521,25 @@ func (r *crdHandler) createCustomResourceDefinition(obj interface{}) {
 	r.customStorageLock.Lock()
 	defer r.customStorageLock.Unlock()
 	// this could happen if the create event is merged from create-update events
-	r.removeStorage_locked(crd.UID)
+	tearDownFinished := r.removeStorage_locked(crd.UID)
+	if utilfeature.DefaultFeatureGate.Enabled(features.StorageVersionAPI) &&
+		utilfeature.DefaultFeatureGate.Enabled(features.APIServerIdentity) {
+		q := r.getOrCreateStorageVersionUpdateQueue_locked(crd.UID)
+		if len(q) == maxStorageVersionUpdateRequests {
+			// When the channel buffer is full, writing to the channel becomes blocking.
+			// Since we still hold the customStorageLock when we write to the channel,
+			// a too-frequently-updated CRD can block the entire API server. Here we
+			// give up updating storage version for this CRD and print a log.
+			// TODO(roycaihw): use warning instead of info when StorageVersionAPI graduates to beta/GA
+			klog.V(4).Infof("Skipping the storage version update for CRD with UID %v due to the queue being full (queue size: %v).",
+				crd.UID, maxStorageVersionUpdateRequests)
+			return
+		}
+		q <- &storageVersionUpdateRequest{
+			crd:  crd,
+			wait: tearDownFinished,
+		}
```

Comment:
Why do we need to update the storage version here? I thought it's enough to do a storageVersionUpdate in `getOrCreateServingInfoFor`.

Is the purpose forcing a storage version update even if there is no write requests? Is that necessary?

Reply:
Note that even if we don't do storage version updates here, we still need to enqueue the req to block later storage version updates until the tearDownFinished is closed.

Reply:
> Is the purpose forcing a storage version update even if there is no write requests?

yes, to update the storage version even if there is no new requests

Reply:
I thought we agreed to only update storage version after there is a write request.

Reply:
I thought differently :) Do you member why we want to do that?

If we only update storage version after there is a new read/write request (note: read requests also trigger underlying storage creation), the correctness of storage version depends on:
1. active CR traffic exists
2. in an HA cluster, the traffic gets routed to all API servers